### PR TITLE
fix(data): rebase stored root offsets when the root is a bundle

### PR DESCRIPTION
**Pull request by arielmelendez**
## Problem

A data item's stored `rootTransactionId` is not guaranteed to be an L1
transaction.

`traverseToRootUsingAttributes` finds the root by walking up the parent chain
and stopping at the first ancestor it holds no attributes for. That rule cannot
distinguish **"the parent is not indexed yet"** from **"this is the L1 root"**.
When it stops early, an intermediate bundle is recorded as the root and
persisted. The pre-computed short-circuit then returns that value forever
without ever validating it, so the record never self-corrects — even once the
parent's own root becomes known.

Chunk retrieval requires an L1 transaction. `TxChunksDataSource.getData` calls
`getTxField(id, 'data_root')` and `getTxOffset(id)`, which for a non-L1 ID can
only fail — but they fail *remotely*, polling up to
`ARWEAVE_PEER_CHUNK_GET_MAX_PEER_ATTEMPT_COUNT` (15) peers first. Measured at
**~90 s per request** before the tier gives up, on a question one local row
lookup answers immediately.

Worked example (public IDs):

```
bzKhHDtGYN_OPL1-ZP0LRvFlM9YtO-YkHDd7duKcVUc   item, 94 B
  └─ stored root: jVtjXfcyjVC4ojENdiUyAhyJo2Ui1QYcQFIZg_zyZEg   <- a data item, not L1
       └─ real L1: WRjLGcmb-_oEkQSET2acVNqTaRV52T6XjLhMAD4jA1g
```

`/tx/jVtjXfcy.../offset` returns `Not Found` on every node — not because the
data is unseeded, but because it is not an L1 transaction.

## The subtlety: the offsets are correct

`rootDataItemOffset` / `rootDataOffset` stored beside a wrong root are **not
garbage**. They are correct *in that root's frame* — a ranged read of
`jVtjXfcy...` at 1409 returns the exact payload. This is what makes the naive
fix dangerous: rewriting the root alone leaves offsets pointing into a different
coordinate system.

The rebase adds the parent's **payload** offset:

```
proposed_root = parent.rootTransactionId
proposed_rdio = parent.rootDataOffset + child.rootDataItemOffset
proposed_rdo  = parent.rootDataOffset + child.rootDataOffset
```

It must be `rootDataOffset` and not `rootDataItemOffset`, because a child's
offsets are measured from the start of the parent bundle's *payload*, not from
the parent's header. Verified: `2997791 + 1409 = 2999200`, and fetching
`WRjLGcmb...` bytes `2999200-2999293` returns the item payload byte-for-byte
from three independent gateways.

## Changes

1. **`rebaseStoredRootIfBundled()`** — validates a stored root with a local
   attribute lookup. If it carries a root of its own it is not L1, so walk up
   and rebase offsets at each hop. Bounded by `MAX_BUNDLE_NESTING_DEPTH` with
   cycle detection. The correction is persisted **only** when the chain resolves
   fully to an L1 transaction, so a still-bundled root is never written back.
2. **`provisional` flag** — the traversal fallback still serves a request using
   a root inferred from an unindexed ancestor, but no longer persists it.
   Storing that guess is what mis-roots items permanently.
3. **`root_tx_stored_root_rebased_total{outcome}`** — `resolved` / `incomplete`,
   so the correction rate is observable after rollout.

## Scale

Across two canary gateways: **102,353** and **102,431** rows are provably
mis-rooted (the stored root itself carries a root), agreeing to 0.08%. This is a
lower bound; a further ~1.58M rows have an indexed root whose own root is null,
which is ambiguous. 97.7% of the confirmed rows descend from a single large
nested bundle. Current accrual is ~2 rows/day.

## Test plan

- 6 new unit tests: rebase onto L1, persistence of the correction, no-op on a
  correctly rooted item, multi-level nesting, refusal to rebase when the parent
  has no payload offset, cycle handling.
- One existing assertion updated: the pre-computed path now performs 2 attribute
  reads instead of 1 (the item's own, plus the root validation). Both local.
- `root-parent-data-source.test.ts`: 46/46 pass. `yarn lint:check` clean.

## Risk to watch

The guard adds an attribute lookup to the pre-computed path. It is LRU-cached in
`CompositeDataAttributesSource`, so normally free, but a cache miss becomes a
round trip to the SQLite worker. On a loaded node that worker has been observed
taking a long time to resolve a miss, so this could slow the common path while
fixing the rare one. Canary rollout will compare median latency on
correctly-rooted traffic, not only the success rate on mis-rooted items. If it
regresses, the validation needs a cheaper gate in front of it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/data/root-parent-data-source.ts [hunk 1]
```diff
@@ -202,6 +202,130 @@ export class RootParentDataSource implements ContiguousDataSource {
...
+
+      totalOffset += rootAttributes.rootDataOffset;
+      rootDataOffset += rootAttributes.rootDataOffset;
+      rootTxId = parentRootTxId;
+      hops += 1;
+    }
+
+    if (hops === 0) {
+      return { ...stored, fromPreComputed: true };
+    }
```
[the remark this answers is not in the document]
Comment (arielmelendez) [reply src/data/root-parent-data-source.ts:345]: Confirmed and fixed in `ceacec28`.

You're right about the ordering: the `visited` check runs at the top of the loop, after the previous iteration already advanced `rootTxId` and both offsets. A chain cycling back to the data item returned the accumulated offsets paired with the item's own ID, and `getData` then fetched that ID with a bundle-frame region.

Now tracked via a `cycleDetected` flag; on a cycle the partial rebase is discarded and the stored pair is returned unchanged, matching the missing-payload-offset branch.

Worth calling out that the unit test asserted the buggy output (`dataCall.id === ITEM`, offset 1419), so it codified the defect rather than catching it. It now asserts `MID` / 1409, that nothing is persisted, and that the `incomplete` counter increments.

**Comment by arielmelendez** [comment]
Both nitpicks addressed in `ceacec28` as well.

**Counter missed first-hop failures** — correct, and it mattered more than a nitpick: the two exits at `hops === 0` (ancestor with no payload offset, failed attribute read) are precisely the "mis-rooted rows are still being read" condition the help text describes, so the metric under-reported exactly the cases worth alerting on. Outcomes are now `resolved` / `incomplete` / `lookup_failed`. I split `lookup_failed` out rather than folding it into `incomplete` because a failed read proves nothing about the root — it means the check itself did not run. Increments are gated on bundling having been confirmed, so correctly rooted items never touch the counter.

**No test coverage for the counter** — added a `readRebased` helper alongside `readLocalResolve` and asserted deltas: `resolved` on the rebase test, `incomplete` on both the missing-payload-offset and cycle tests, and zero on all three labels for the correctly-rooted case.

46/46 tests pass, `yarn lint:check` clean.

**Comment by arielmelendez** [comment]
## Canary results

Deployed to one canary gateway. Two disjoint 25-item groups per phase, all on the same host, measured serially.

**Mis-rooted items (the target):**

| | Success | Median | Max |
|---|---|---|---|
| before | 2/25 — 8% | 6.3 s | 163.8 s |
| after | **24/25 — 96%** | 11.1 s | 82.0 s |

**Correctly rooted items (regression control):**

| | Success | Median | p90 | Max |
|---|---|---|---|---|
| before | 23/25 — 92% | 1.97 s | 17.1 s | 51.3 s |
| after | 25/25 — 100% | 1.07 s | 9.1 s | 11.2 s |

The added attribute lookup did not regress the common path — the risk flagged in the description did not materialise. Median on mis-rooted items rises because failures used to 404 quickly; the max drops as the ~90 s chunk dead-end stops being entered.

End-to-end on the item that started this, with caches cold on both hosts:

```
Rebased stored root onto its L1 transaction
  dataItemId: bzKhHDtGYN_OPL1-ZP0LRvFlM9YtO-YkHDd7duKcVUc
  hops: 1  outcome: resolved
  rebasedRootTxId: WRjLGcmb-_oEkQSET2acVNqTaRV52T6XjLhMAD4jA1g
  rebasedRootDataOffset: 2999200
```

2999200 matches the value computed by hand and verified against three independent gateways. 180 s → 7.8 s.

## Correction pushed in `fe74861d`

While verifying the above I found the description and the metric help overstated the fix: **the correction is not persisted.** `CompositeDataAttributesSource.setDataAttributes` writes to its in-process LRU and never calls through to the database. Proof: after the rebase above, `contiguous_data_ids` still held the intermediate bundle and the original offsets.

Serving is fixed, which is the point. But rows are not repaired, the counter does not decay as items are read, and a backfill is still required to fix stored state. TSDoc and the counter help now say so.

**Comment by arielmelendez** [comment]
## Correction to my previous comment — self-healing does work

I reported that a rebased root is never persisted. That was wrong, and the evidence was an artefact of my test.

`setDataAttributes` is LRU-only, but `ReadThroughDataCache` re-reads attributes after a successful retrieval and queues them through `DataContentAttributeImporter`; `insertDataId` then overwrites the root columns via `excluded.*`, gated only on `verified != 1`.

What blocked it was `StandaloneSqlite`'s `saveDataContentAttributes` dedupe — an LRU keyed on ID with a **7-minute TTL**. My two cold fetches were 65 seconds apart. The first resolved via a pre-guard tier, re-persisted the existing values, and claimed the slot; the guard's correction was dropped.

Repeating the fetch outside that window:

```
root = 5918CB19…  (WRjLGcmb, the L1 root)
rdio = 2997951    rdo = 2999200    verified = 1
```

The row repaired itself. Corrected in `3c1f151d` (TSDoc + counter help).

Two things worth recording:

- A cache-fill re-persists root attributes even when unchanged — `standalone_sqlite_method_duplicate_calls_total{method="saveDataContentAttributes"}` was 346 in ~40 min. Harmless but wasteful on the write path; keying the dedupe on `id + rootTransactionId` would fix both that and the dropped-correction case.
- The upsert is gated on `verified != 1`, so a **verified** row can never be corrected by this path. Currently 0 of the ~102k affected rows are verified, so there is no exposure today — but if verification ever runs ahead of repair, those rows become backfill-only.

File: src/database/standalone-sqlite.ts [hunk 2]
```diff
@@ -4137,14 +4137,33 @@ export class StandaloneSqliteDatabase
...
+    // before RootParentDataSource would re-persist the item's existing root,
+    // claim the slot, and a corrected root from the rebase walk arriving inside
+    // the same window was silently dropped. Including the root fields keeps
+    // suppressing writes that would change nothing, while letting a write that
+    // actually moves the item's root coordinates through every time.
+    //
+    // This can only admit writes the ID-only key would have suppressed, never
+    // suppress ones it allowed, and the extra writes are limited to genuine
+    // corrections — so the protection this cache exists to give the write
+    // queue is preserved.
```
[the remark this answers is not in the document]
Comment (arielmelendez) [reply src/database/standalone-sqlite.ts:4166]: Fixed in `db15fcdb`.

File: src/database/standalone-sqlite.ts [hunk 3]
```diff
@@ -4137,14 +4137,33 @@ export class StandaloneSqliteDatabase
...
+    const dedupeKey = [
+      id,
+      rootTransactionId ?? '',
+      rootDataItemOffset ?? '',
+      rootDataOffset ?? '',
+    ].join('|');
```
[the remark this answers is not in the document]
Comment (arielmelendez) [reply src/database/standalone-sqlite.ts:4172]: Fixed in `db15fcdb`.

**Comment by arielmelendez** [comment]
Both comments addressed in `db15fcdb`.

**Cache invalidation for composite keys** — confirmed, and this was a real regression I introduced. Keying the dedupe on root coordinates left `clearDataHash()` deleting a bare ID that is no longer a key, so it invalidated nothing. Concretely: after a cache re-verification mismatch evicted a blob, the re-save that must follow was suppressed for the rest of the 7-minute TTL, stranding the row with a null hash.

Fixed by dropping every entry belonging to the item. A prefix scan rather than a secondary index because the only caller is the re-verification mismatch path in `ReadThroughDataCache` — rare and exceptional — and the cache is bounded at `DEDUPE_CACHE_MAX_SIZE` (10k). `|` cannot appear in a base64url ID, so the prefix is unambiguous.

The regression test fails against the bare-ID delete and passes with the scan, so it tests the real thing.

**TSDoc** — added to both touched methods, documenting the dedupe contract and, on `clearDataHash`, the invalidation requirement that links them.

One thing worth recording for anyone writing tests here: my first version of the regression test passed alone and failed in a full run. `insertDataHashCache` is an in-process LRU that outlives the per-test database reset, so reusing a hash literal another test already wrote makes `insertDataHash` a no-op and the `contiguous_data` row never appears. The test now uses a hash unique to itself, with a comment explaining why.

71/71 in `standalone-sqlite`, 46/46 in `root-parent-data-source`, lint and typecheck clean.

File: src/database/standalone-sqlite.test.ts [hunk 4]
```diff
@@ -2143,6 +2143,104 @@ describe('StandaloneSqliteDatabase', () => {
...
+      // The re-save carries identical root coordinates and lands inside the
+      // dedupe TTL. Because the keys embed those coordinates, clearing by bare
+      // ID would miss them and this write would be suppressed, stranding the
+      // row with a null hash until the window expired.
+      await db.saveDataContentAttributes(attrs);
+      assert.notEqual(
+        (await db.getDataAttributes(EVICTED))?.hash,
+        undefined,
+        'hash must be restored: the dedupe entry should have been invalidated',
+      );
```
[the remark this answers is not in the document]
Comment (arielmelendez) [reply src/database/standalone-sqlite.test.ts:2237]: Good catch — fixed in `9da3d267`. You're right that `notEqual(…, undefined)` would have accepted a *different* hash, which is arguably the worse outcome: the row rewritten with wrong content rather than left null. Now captures the hash from the initial save and asserts equality against it.

Re-checked that the test still fails against the bare-ID delete and passes with the prefix scan, so it continues to cover the regression it was written for.

(Edited to correct the commit hash — the original reply cited a hash that did not exist.)
